File: app/src/main/java/hegde/mahesh/avre/ui/home/HomeViewModel.java
package hegde.mahesh.avre.ui.home;

import android.app.Application;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import hegde.mahesh.avre.database.AvreDatabase;
import hegde.mahesh.avre.database.HistoryRepository;
import hegde.mahesh.avre.interpreter.BshInterpreter;
import hegde.mahesh.avre.interpreter.LanguageInterpreter;
import hegde.mahesh.avre.interpreter.SnippetEvalException;
import hegde.mahesh.avre.model.HistoryItem;

import java.io.ByteArrayOutputStream;
import java.util.List;

/// Stores HomeFragment state
public class HomeViewModel extends AndroidViewModel {
    private final LanguageInterpreter interpreter = new BshInterpreter();
    public LiveData<String> getCurrentSnippet() {
        return currentSnippet;
    }

    private final MutableLiveData<String> currentSnippet = new MutableLiveData<>("");

    public LiveData<Integer> getVisibleHistoryStart() {
        return visibleHistoryStart;
    }

    public final MutableLiveData<Integer> visibleHistoryStart = new MutableLiveData<>(0);

    public LiveData<Integer> getLastAppendedPosition() {
        return lastAppendedPosition;
    }

    public final MutableLiveData<Integer> lastAppendedPosition = new MutableLiveData<>(-1);

    private int cursor;

    private final HistoryRepository repo;

    public List<HistoryItem> getHistoryItems() {
        return repo.getHistory();
    }
    private final ByteArrayOutputStream out = new ByteArrayOutputStream();
    private final ByteArrayOutputStream err = new ByteArrayOutputStream();

    public HomeViewModel(Application application) {
        super(application);
        AvreDatabase db = AvreDatabase.getDatabase(application);
        repo = new HistoryRepository(db);
        interpreter.setOutputStream(out);
        interpreter.setErrorStream(err);
    }

    public void eval(String snippet) {
        try {
            Object result = interpreter.eval(snippet);
            String valClass = result != null ? (": " + result.getClass()) : "";

            String outStr = readOutput(out);
            String errStr = readOutput(err);

            HistoryItem historyItem = HistoryItem.success(snippet, outStr, errStr,
                    result + valClass);
            repo.addItem(historyItem);
        } catch (Exception e) {
            HistoryItem failure = HistoryItem.failure(snippet, readOutput(out), readOutput(err), e);
            repo.addItem(failure);
        }
        cursor = repo.getHistorySize();
        lastAppendedPosition.setValue(cursor - 1);
    }

    public void setInterpreterVariable(String name, Object value) {
        try {
            interpreter.setVariable(name, value);
        } catch (SnippetEvalException e) {
            throw new RuntimeException(e);
        }
    }

    public void clearVisibleHistory() {
        visibleHistoryStart.setValue(repo.getHistorySize());
    }

    public void reset() {
        repo.reset();
        cursor = 0;
        visibleHistoryStart.setValue(0);
        interpreter.reset();
    }

    public void previousSnippet() {
        String code;
        List<HistoryItem> history = repo.getHistory();
        if (cursor - 1 >= 0) {
            cursor = cursor - 1;
            code = history.get(cursor).code;
        } else if (!history.isEmpty()) {
            code = history.get(0).code;
        } else {
            code = "";
        }
        currentSnippet.setValue(code);
    }

    public void nextSnippet() {
        List<HistoryItem> history = repo.getHistory();
        String code;
        cursor = Math.min(cursor + 1, history.size());
        if (cursor == history.size()) {
            code = "";
        } else {
            code = history.get(cursor).code;
        }
        currentSnippet.setValue(code);
    }

    private static String readOutput(ByteArrayOutputStream stream) {
        String result = stream.toString();
        stream.reset();
        return result;
    }
}

File: app/src/main/java/hegde/mahesh/avre/ui/home/HomeViewModel.kt
package hegde.mahesh.avre.ui.home

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import hegde.mahesh.avre.database.AvreDatabase.Companion.getDatabase
import hegde.mahesh.avre.database.HistoryRepository
import hegde.mahesh.avre.interpreter.BshInterpreter
import hegde.mahesh.avre.interpreter.LanguageInterpreter
import hegde.mahesh.avre.interpreter.SnippetEvalException
import hegde.mahesh.avre.model.HistoryItem
import hegde.mahesh.avre.model.HistoryItem.Companion.failure
import hegde.mahesh.avre.model.HistoryItem.Companion.success
import java.io.ByteArrayOutputStream

/// Stores HomeFragment state
class HomeViewModel(application: Application) : AndroidViewModel(application) {
    private val interpreter: LanguageInterpreter = BshInterpreter()

    private val mCurrentSnippet = MutableLiveData("")
    val currentSnippet: LiveData<String>
        get() = mCurrentSnippet

    private val mVisibleHistoryStart = MutableLiveData(0)
    val visibleHistoryStart: LiveData<Int>
        get() = mVisibleHistoryStart

    private val mHistorySize = MutableLiveData<Int>()
    val historySize: LiveData<Int>
        get() = mHistorySize

    private var cursor = 0

    private val repo: HistoryRepository

    val historyItems: List<HistoryItem>
        get() = repo.history

    private val out = ByteArrayOutputStream()
    private val err = ByteArrayOutputStream()

    init {
        val db = getDatabase(application)
        repo = HistoryRepository(db)
        initializeInterpreter()
    }

    private fun initializeInterpreter() {
        interpreter.setOutputStream(out)
        interpreter.setErrorStream(err)
    }

    fun eval(snippet: String) {
        try {
            val result = interpreter.eval(snippet)
            val valClass = if (result != null) ": " + result.javaClass else ""
            val outStr = readOutput(out)
            val errStr = readOutput(err)
            val historyItem = success(
                snippet, outStr, errStr,
                result.toString() + valClass
            )
            repo.addItem(historyItem)
        } catch (e: Exception) {
            val failure = failure(snippet, readOutput(out), readOutput(err), e)
            repo.addItem(failure)
        }
        val size = repo.getHistorySize()
        cursor = size
        mHistorySize.value = size
    }

    fun setInterpreterVariable(name: String, value: Any?) {
        try {
            interpreter.setVariable(name, value)
        } catch (e: SnippetEvalException) {
            throw RuntimeException(e)
        }
    }

    fun clearVisibleHistory() {
        mVisibleHistoryStart.value = repo.getHistorySize()
        mHistorySize.value = 0
        cursor = 0
    }

    fun reset() {
        repo.reset()
        interpreter.reset()
        clearVisibleHistory()
    }

    fun previousSnippet() {
        val code: String
        val history: List<HistoryItem> = repo.history
        if (cursor - 1 >= 0) {
            cursor -= 1
            code = history[cursor].code
        } else if (history.isNotEmpty()) {
            code = history[0].code
        } else {
            code = ""
        }
        mCurrentSnippet.value = code
    }

    fun nextSnippet() {
        val history: List<HistoryItem> = repo.history
        cursor = (cursor + 1).coerceAtMost(history.size)
        val code: String = if (cursor == history.size) {
            ""
        } else {
            history[cursor].code
        }
        mCurrentSnippet.value = code
    }

    companion object {
        private fun readOutput(stream: ByteArrayOutputStream): String {
            val result = stream.toString()
            stream.reset()
            return result
        }
    }
}